kml/serdes_gpx.cpp

```diff
@@ -0,0 +1,273 @@
+  }
+  
+}
+
+std::string const & GpxParser::GetTagFromEnd(size_t n) const
+{
+  ASSERT_LESS(n, m_tags.size(), ());
+  return m_tags[m_tags.size() - n - 1];
+}
+
+void GpxParser::Pop(std::string const & tag)
+{
+  ASSERT_EQUAL(m_tags.back(), tag, ());
+  
+  if (tag == "trkpt")
+  {
+    m2::Point p = mercator::FromLatLon(m_lat, m_lon);
+    
+    if (m_line.empty() || !AlmostEqualAbs(m_line.back().GetPoint(), p, kMwmPointAccuracy))
+      m_line.emplace_back(p);
+    
+  }
+  else if (tag == "trkseg")
+  {
+    m_geometry.m_lines.push_back(std::move(m_line));
+  }
+  else if (tag == "wpt")
+  {
+    m_org = mercator::FromLatLon(m_lat, m_lon);
+  }
+  
+  if (tag == "trkseg" || tag == "wpt")
+  {
+    if (MakeValid())
+    {
+      if (GEOMETRY_TYPE_POINT == m_geometryType)
+      {
+        BookmarkData data;
+        data.m_name = std::move(m_name);
+        data.m_description = std::move(m_description);
+        data.m_color.m_predefinedColor = m_predefinedColor;
+        data.m_color.m_rgba = m_color;
+        data.m_icon = m_icon;
+        data.m_viewportScale = m_viewportScale;
+        data.m_timestamp = m_timestamp;
+        data.m_point = m_org;
+        data.m_featureTypes = std::move(m_featureTypes);
+        data.m_customName = std::move(m_customName);
+        data.m_boundTracks = std::move(m_boundTracks);
+        data.m_visible = m_visible;
+        data.m_nearestToponym = std::move(m_nearestToponym);
+        data.m_minZoom = m_minZoom;
+        data.m_properties = std::move(m_properties);
+        data.m_compilations = std::move(m_compilations);
+        
+        // Here we set custom name from 'name' field for KML-files exported from 3rd-party services.
+        if (data.m_name.size() == 1 && data.m_name.begin()->first == kDefaultLangCode &&
+            data.m_customName.empty() && data.m_featureTypes.empty())
+        {
+          data.m_customName = data.m_name;
+        }
+        
+        m_data.m_bookmarksData.push_back(std::move(data));
+      }
+      else if (GEOMETRY_TYPE_LINE == m_geometryType)
+      {
+        TrackData data;
+        data.m_localId = m_localId;
+        data.m_name = std::move(m_name);
+        data.m_description = std::move(m_description);
+        data.m_layers = std::move(m_trackLayers);
+        data.m_timestamp = m_timestamp;
+        data.m_geometry = std::move(m_geometry);
+        data.m_visible = m_visible;
+        data.m_nearestToponyms = std::move(m_nearestToponyms);
+        data.m_properties = std::move(m_properties);
+        m_data.m_tracksData.push_back(std::move(data));
+      }
+    }
+    ResetPoint();
+  }
+  m_tags.pop_back();
+}
+
+void GpxParser::CharData(std::string value)
+{
+  strings::Trim(value);
+  
+  size_t const count = m_tags.size();
+  if (count > 1 && !value.empty())
+  {
+    std::string const & currTag = m_tags[count - 1];
+    std::string const & prevTag = m_tags[count - 2];
+    std::string const ppTag = count > 2 ? m_tags[count - 3] : std::string();
```

Comment:
Do you really need these tags? What does linter say?

Reply:
Looks like copy-paste from KML where they are used for some specific case and not used in GPX for now, so removed.